### PR TITLE
SI-7747 Make REPL wrappers serialization friendly

### PR DESCRIPTION
**Pull request by retronym**
Review by @ScrapCodes / @som-snytt

Minor reworking of #4493.

### CODE REVIEW

**Comment by ScrapCodes** [comment]
Thanks @retronym, I am going to try out the effects of doing it only for val handlers. Please allow me some time.

**Comment by retronym** [comment]
Thanks! I restricted it to that to make sure the code was test driven. If you find something that misses, we can add it to the test.

**Comment by retronym** [comment]
Failed test:

```
% diff /home/jenkins/workspace/scala-2.11.x-validate-test/test/files/run/repl-serialization-run.log /home/jenkins/workspace/scala-2.11.x-validate-test/test/files/run/repl-serialization.check
@@ -16,6 +16,6 @@ defined type alias AA
   evaluating O
   constructing A
 == reconstituting into a fresh classloader
-== evaluating reconstituted lambda
   evaluating O
+== evaluating reconstituted lambda
   constructing A
```

Just pushed a new version of the second commit that treats modules and vals in the same way. That makes the test expectation consistent with the first commit.

**Comment by ScrapCodes** [comment]
Did you also notice the difference for excluding ModuleHandler ?

**Comment by retronym** [comment]
The test added in the first commit failed in the original version of the second commit, as shown in that diff. The module `O`, defined on a previous line, was initialized upon deserialization of the closure, whereas before it was initialized when it was accessed during execution of that closure.

In both cases, the module will be initialized on both the client and server, it is is just a matter of how soon.

**Comment by ScrapCodes** [comment]
I have one test that fails

``` scala

scala> case class Sum(exp: String, exp2: String)
defined class Sum

scala> val a = Sum("A", "B")
a: Sum = Sum(A,B)

scala> def b(a: Sum): String = a match { case Sum(_, _) => "Found Sum" }
b: (a: Sum)String

scala> b(a)
<console>:15: error: type mismatch;
 found   : Sum
 required: Sum
              b(a)
```

And there are more, but I am just trying to understand why they are failing. 

EDIT: It failed only at first commit not at second. I just forgot that when I was testing.

**Comment by ScrapCodes** [comment]
If you apply this change to your test case.

``` diff
diff --git a/test/files/run/repl-serialization.scala b/test/files/run/repl-serialization.scala
index 64915ce..9460c5d 100644
--- a/test/files/run/repl-serialization.scala
+++ b/test/files/run/repl-serialization.scala
@@ -24,13 +24,14 @@ object Test {
     }

     val code =
-      """val x = {println("  evaluating x"); 0 }
+      """var x = {println("  evaluating x"); 0 }
+        |def getX() = x
         |lazy val y = {println("  evaluating y"); 0 }
         |class D; val z = {println("  evaluating z"); 0}; val zz = {println("  evaluating zz"); 0}
         |object O extends Serializable { val apply = {println("  evaluating O"); 0} }
         |class A(i: Int) { println("  constructing A") }
         |type AA = A
-        |extract(() => new AA(x + y + z + zz + O.apply))
+        |extract(() => new AA(x + getX() + y + z + zz + O.apply))
       """.stripMargin

     imain = new IMain(settings)
```

It results in 

``` diff
diff --git a/test/files/run/repl-serialization.check b/test/files/run/repl-serialization.check
index c8439ea..1c2ba25 100644
--- a/test/files/run/repl-serialization.check
+++ b/test/files/run/repl-serialization.check
@@ -2,6 +2,7 @@
 extract: AnyRef => Unit = <function1>
   evaluating x
 x: Int = 0
+getX: ()Int
 y: Int = <lazy>
   evaluating z
   evaluating zz
@@ -18,4 +19,5 @@ defined type alias AA
 == reconstituting into a fresh classloader
   evaluating O
 == evaluating reconstituted lambda
+  evaluating x
   constructing A
```

This will lead to "evaluating x" again.

To fix this test we can apply this patch.

``` diff
diff --git a/src/repl/scala/tools/nsc/interpreter/Imports.scala b/src/repl/scala/tools/nsc/interpreter/Imports.scala
index c52869e..5030bea 100644
--- a/src/repl/scala/tools/nsc/interpreter/Imports.scala
+++ b/src/repl/scala/tools/nsc/interpreter/Imports.scala
@@ -166,7 +166,7 @@ trait Imports {
             for (imv <- x.definedNames) {
               if (!currentImps.contains(imv)) {
                 x match {
-                  case _: ValHandler | _: ModuleHandler =>
+                  case _: ValHandler | _: ModuleHandler | _: DefHandler=>
                     val valName = req.lineRep.packageName + req.lineRep.readName
                     if (!tempValLines.contains(req.lineRep.lineId)) {
                       code.append(s"val $valName = $objName\n")

```

And the extra "evaluating x" disappears.

**Comment by ScrapCodes** [comment]
@retronym Can you update the test case with above commented patch ? 
It is good to go after that.

**Comment by retronym** [comment]
What about:

```
scala> class C { println(x) }
scala> new C()
```

So I'm wondering if the right answer would be to drop the second commit and uniformly use the temporary  val? WDYT?

**Comment by retronym** [comment]
I just read:

> EDIT: It failed only at first commit not at second. I just forgot that when I was testing.

So we can't just drop the second commit. I wasn't really sure of the intent of the second, as it wasn't supported by my test case. I'll incorporate your two additional tests tomorrow, in addition to a reference to the val from a class, and see where that leaves us.

This stuff is really subtle, so I want to be sure it is backed up by both tests and a solid explanation. We're getting very close though!

**Comment by ScrapCodes** [comment]
Yes, that sounds good to me.

File: src/repl/scala/tools/nsc/interpreter/Imports.scala [hunk 1]
```diff
@@ -157,6 +162,23 @@ trait Imports {
...
+            for (imv <- x.definedNames) {
+              if (!currentImps.contains(imv)) {
+                x match {
+                  case _: ValHandler | _: ModuleHandler =>
+                    val valName = req.lineRep.packageName + req.lineRep.readName
+                    if (!tempValLines.contains(req.lineRep.lineId)) {
+                      code.append(s"val $valName = $objName\n")
+                      tempValLines += req.lineRep.lineId
+                    }
+                    code.append(s"import $valName ${req.accessPath}.`$imv`;\n")
```
Comment (ScrapCodes) [inline src/repl/scala/tools/nsc/interpreter/Imports.scala:175]: I think this should be 

``` scala
code.append(s"import $valName${req.accessPath}.`$imv`;\n")
```

**Comment by ScrapCodes** [comment]
@retronym Did you get a chance to look at it ? This seemed close to finish, is there anything I could do ?

**Comment by retronym** [comment]
I haven't gotten back to this one. I'll be a bit busy for the next week with Scala Days. Would you be able to pick this one up again? Hopefully it just needs the test case expanded, and the use of the `INSTANCE` import made unconditional.

**Comment by ScrapCodes** [comment]
I will look into it and be back with another PR hopefully.

**Pull request closed without merging**